spec/index.md

```diff
@@ -1298,78 +1298,80 @@ In the reply callback of a [inter-canister method call](#system-api-call), the a
-    ic0.call_cycles_add : (amount : i64) -> ();                                 // U Ry Rt T
-    ic0.call_cycles_add128 : (amount_high : i64, amount_low: i64) -> ();        // U Ry Rt T
-    ic0.call_perform : () -> ( err_code : i32 );                                // U CQ Ry Rt CRy CRt T
-
-    ic0.stable_size : () -> (page_count : i32);                                 // * s
-    ic0.stable_grow : (new_pages : i32) -> (old_page_count : i32);              // * s
-    ic0.stable_write : (offset : i32, src : i32, size : i32) -> ();             // * s
-    ic0.stable_read : (dst : i32, offset : i32, size : i32) -> ();              // * s
-    ic0.stable64_size : () -> (page_count : i64);                               // * s
-    ic0.stable64_grow : (new_pages : i64) -> (old_page_count : i64);            // * s
-    ic0.stable64_write : (offset : i64, src : i64, size : i64) -> ();           // * s
-    ic0.stable64_read : (dst : i64, offset : i64, size : i64) -> ();            // * s
-
-    ic0.certified_data_set : (src: i32, size: i32) -> ();                       // I G U Ry Rt T
-    ic0.data_certificate_present : () -> i32;                                   // *
-    ic0.data_certificate_size : () -> i32;                                      // Q CQ
-    ic0.data_certificate_copy : (dst: i32, offset: i32, size: i32) -> ();       // Q CQ
-
-    ic0.time : () -> (timestamp : i64);                                         // *
-    ic0.global_timer_set : (timestamp : i64) -> i64;                            // I G U Ry Rt C T
-    ic0.performance_counter : (counter_type : i32) -> (counter : i64);          // * s
-    ic0.is_controller: (src: i32, size: i32) -> ( result: i32);                 // * s
-    ic0.in_replicated_execution: () -> (result: i32);                           // * s
-
-    ic0.debug_print : (src : i32, size : i32) -> ();                            // * s
-    ic0.trap : (src : i32, size : i32) -> ();                                   // * s
+      ) -> (); I ∈ {i32, i64}                                                             // U CQ Ry Rt CRy CRt T
+    ic0.call_on_cleanup : (fun : i32, env : i32) -> ();                                   // U CQ Ry Rt CRy CRt T
+    ic0.call_data_append : (src : I, size : i32) -> (); I ∈ {i32, i64}                    // U CQ Ry Rt CRy CRt T
+    ic0.call_cycles_add : (amount : i64) -> ();                                           // U Ry Rt T
+    ic0.call_cycles_add128 : (amount_high : i64, amount_low: i64) -> ();                  // U Ry Rt T
+    ic0.call_perform : () -> ( err_code : i32 );                                          // U CQ Ry Rt CRy CRt T
+
+    ic0.stable_size : () -> (page_count : i32);                                           // * s
+    ic0.stable_grow : (new_pages : i32) -> (old_page_count : i32);                        // * s
+    ic0.stable_write : (offset : i32, src : i32, size : i32) -> ();                       // * s
```

Comment:
Conceptually, `src` would also become `I` for `stable_write` and `stable64_write` (distinguishing the address space of main memory from stable memory). Analogously for `dst` in `stable_read` and `stable64_read`. However, I also see good rationale to simply keep the current two explicit versions for historical reasons.

Reply:
I kept the types here since we already have the two versions with separate names and I don't think we could ever deprecate one of the two versions.